server/aesdsocket.c

```diff
@@ -11,144 +21,245 @@
     server.sin_port = htons(SERVER_PORT);
     server.sin_addr.s_addr = htonl(INADDR_ANY);
-    
     if (bind(sockfd, (struct sockaddr *)&server, sizeof(server)) < 0) {
         perror("bind: Failed to bind name to socket.");
         close(sockfd);
-        return -1;
+        exit(-1);
     }
 
+    // Listen for client connections.
     if (listen(sockfd, 5) < 0) {
         perror("listen: Failed to listen for connections.");
         close(sockfd);
-        return -1;
+        exit(-1);
     }
 
+    // Main loop to accept client connections.
     while (1) {
         struct sockaddr_in client;
         socklen_t client_len = sizeof(client);
+        int *new_sock = malloc(sizeof(int));
         int connfd = accept(sockfd, (struct sockaddr*)&client, &client_len);
-
         if (connfd < 0) {
             perror("accept: Failed connecting to client.");
+            free(new_sock);
             continue;
         }
-        
+
+        // Log the client IP address.
         struct sockaddr_in addr;
         socklen_t addr_len = sizeof(addr);
         getpeername(connfd, (struct sockaddr *)&addr, &addr_len);
         char *client_ip = inet_ntoa(addr.sin_addr);
-        printf("Accepted connection from %s\n", client_ip);
         syslog(LOG_INFO, "Accepted connection from %s", client_ip);
         
-        char* buffer = calloc(BUFFER_SIZE, sizeof(char) * BUFFER_SIZE);
-
-        while (recv(connfd, buffer, sizeof(buffer), 0) != 0) {
-            FILE *fp = fopen("/var/tmp/aesdsocketdata", "a");
-            if (fp == NULL) {
-                printf("fopen: Failed opening file.\n");
-                perror("fopen: Failed opening file.");
-                return -1;
-            } else {
-                fprintf(fp, "%s", buffer);
-                fclose(fp);
-            }
+        *new_sock = connfd;
 
-            if (strchr(buffer, '\n') != NULL) {
-                FILE *fp_out = fopen("/var/tmp/aesdsocketdata", "r");
-                if (fp_out == NULL) {
-                    printf("fopen: Failed opening file.\n");
-                    perror("fopen: Failed opening file.");
-                    return -1;
-                } else {
-                    uint32_t bytes_read = 0;
-                    while ((bytes_read = fread(buffer, 1, sizeof(buffer), fp_out)) != 0) {
-                        send(connfd, buffer, bytes_read, 0);
-                    }                
-                }
-                fclose(fp_out);
-            }
-            memset(buffer, 0, sizeof(char) * BUFFER_SIZE);
+        pthread_t new_thread;
+        if (pthread_create(&new_thread, NULL, connection_handler, new_sock) != 0) {
+            perror("pthread_create failed");
+            close(connfd);
+            free(new_sock);
+            continue;
         }
-        free(buffer);
-        close(connfd);
-        syslog(LOG_INFO, "Closed connection from %s", client_ip);
+
+        add_thread(new_thread);
 
     }
+        
+    node_t *current = head;
```

Comment:
@tiba6275 I don't think you are hitting this cleanup code due to the `exit` in your signal handler.  This is related to these valgrind warnings:
```
==919== 4,624 bytes in 17 blocks are possibly lost in loss record 4 of 4
==919==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==919==    by 0x40149CA: allocate_dtv (dl-tls.c:286)
==919==    by 0x40149CA: _dl_allocate_tls (dl-tls.c:532)
==919==    by 0x4874322: allocate_stack (allocatestack.c:622)
==919==    by 0x4874322: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
==919==    by 0x10AE5A: main (aesdsocket.c:238)
```